src/Feedback/FB_AdvanceDt.cpp

```diff
@@ -0,0 +1,472 @@
+                  for (int p=0; p<NPar; p++)
+                     ParAtt_Local[v][p] = amr->Par->Attribute[v][ ParList[p] ];
+            }
+         } // if ( UseParAttCopy ) ... else ...
+
+
+//       7-3. sort particles by positions to fix their order
+//            --> necessary when feedback involves random numbers
+//            --> otherwise, the same particles accessed by different patches may have different random numbers
+         Par_SortByPos( NPar, ParAtt_Local[PAR_POSX], ParAtt_Local[PAR_POSY], ParAtt_Local[PAR_POSZ], ParSortID );
+
+
+
+//       7-4. periodic boundary conditions
+//            --> assuming there are at least TWO patch groups along each spatial direction (i.e., NX0_TOT[*]/PS2>=2)
+//            --> so each particle won't affect the same patch group more than once
+         const bool   Periodic [3] = { OPT__BC_FLU[0] == BC_FLU_PERIODIC,
+                                       OPT__BC_FLU[2] == BC_FLU_PERIODIC,
+                                       OPT__BC_FLU[4] == BC_FLU_PERIODIC };
+         const double HalfBox  [3] = { 0.5*amr->BoxSize[0], 0.5*amr->BoxSize[1], 0.5*amr->BoxSize[2] };
+         const int    ParPosIdx[3] = { PAR_POSX, PAR_POSY, PAR_POSZ };
+
+         for (int d=0; d<3; d++) {
+            if ( Periodic[d] ) {
+               for (int p=0; p<NPar; p++) {
+
+                  real *ParPos = ParAtt_Local[ ParPosIdx[d] ] + p;
+                  const double dr = *ParPos - PGCenter[d];
+
+                  if      ( dr > +HalfBox[d] )  *ParPos -= amr->BoxSize[d];
+                  else if ( dr < -HalfBox[d] )  *ParPos += amr->BoxSize[d];
+               }
+            } // if ( Periodic[d] )
+         } // for (int d=0; d<3; d++)
+
+
+
+//       8. feedback
+//       8-1. set the random seed
+//            --> to get deterministic and different random numbers for all patches, reset the random seed of
+//                each patch according to its location and counter
+//            --> factors 1e2 and 1e8 are to make random seeds more different
+         const long RSeed = FB_RSEED + amr->patch[0][lv][PID]->LB_Idx*100L + AdvanceCounter[lv]*100000000L;
+         FB_RNG->SetSeed( TID, RSeed );
+
+
+//       8-2. invoke all feedback routines
+         const double EdgeL[3] = { amr->patch[0][lv][PID0]->EdgeL[0] - FB_GHOST_SIZE*amr->dh[lv],
+                                   amr->patch[0][lv][PID0]->EdgeL[1] - FB_GHOST_SIZE*amr->dh[lv],
+                                   amr->patch[0][lv][PID0]->EdgeL[2] - FB_GHOST_SIZE*amr->dh[lv] };
+         int Status;
+
+         if ( FB_SNE  )    Status = FB_SNe     ( lv, TimeNew, TimeOld, dt, NPar, ParSortID, ParAtt_Local, fluid_PG,
+                                                 EdgeL, amr->dh[lv], CoarseFine, TID, FB_RNG );
+
+         if ( FB_USER )    Status = FB_User_Ptr( lv, TimeNew, TimeOld, dt, NPar, ParSortID, ParAtt_Local, fluid_PG,
+                                                 EdgeL, amr->dh[lv], CoarseFine, TID, FB_RNG );
+
```

Comment:
Different feedback routines may use the return value `Status` in different ways (i.e., it may not be necessary to be `GAMER_SUCCESS` or `GAMER_FAILED`). So I'll let people implementing these feedback routines to decide how to handle it.